ebill_paynet/components/api.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+import os
+
+from lxml import html
+from requests import Session
+from zeep import Client, Settings
+from zeep.transports import Transport
+
+from odoo.modules.module import get_resource_path
+
+WSDL_DOC = os.path.join(os.path.dirname(__file__), "wsdl", "DWSPayNet.wsdl")
+SSL_PROD_CERTIFICATE = get_resource_path(
+    "ebill_paynet", "certificats", "prod_services_chain.pem"
+)
+SSL_TEST_CERTIFICATE = get_resource_path(
+    "ebill_paynet", "certificats", "test_services_chain.pem"
+)
+
+
+class PayNetDWS:
+    """PayNet DWS web services."""
+
+    def __init__(self, url, test_service):
+        settings = Settings(xml_huge_tree=True)
```

Comment:
Yes clearer, but with an `import zeep` are we not importing more than necessary and it would be bad performance wise ? (real question that I don't know the answer)

Reply:
@TDu good question to ask.

AFAICT imports does 2 things they import the modules and submodules and fill the namespace.

`from a_module import *` is discouraged only because you fill the namespace with every part of  `a_module`.

When using the `from` you target a submodule or code of the main module in any case you will also load the module itself.

If you test the following in your python shell to load :

```
>>> import sys
>>> sys.modules['zeep']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
KeyError: 'zeep'
>>> from zeep import Settin
>>> Settings
<class 'zeep.settings.Settings'>
>>> sys.modules['zeep']
<module 'zeep' from '/home/yvaucher/.local/lib/python3.6/site-packages/zeep/__init__.py'>
>>> sys.modules['zeep'].Settings
<class 'zeep.settings.Settings'>
>>> sys.modules['zeep'].Client
<class 'zeep.client.Client'>
>>> Client
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'Client' is not defined
```

As you can see the `zeep` module is fully loaded but the namespace doesn't contains everything.



But what is a submodule?
https://docs.python.org/3/reference/import.html#submodules

A submodule is a sub file or directory that is not mentionned in the `__init__.py`
It's different from code that is part of the main module which must appear in the `__init__.py` file.



To deep in details, you can't import too much things unless you ask for a submodule you don't use.

Consider the following module with 2 submodules and some initialization code:

```
module
├── initialization.py
├── __init__.py
├── sub1.py
└── sub2.py
```

The content of `module/__init__.py` is 
``` python
from . import initialization

print('import whole module')
```

Now let's test it:

```
>>> import module
load initialization
import whole module
>>> module.initialization
<module 'module.initialization' from '/tmp/module/initialization.py'>
>>> module.sub1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'module' has no attribute 'sub1'
```
Only what is mention as `from . import ....` is loaded with the module.

-----------------
```
>>> from module import initialization
load initialization
import whole module
>>> from module import sub1
load sub1
>>> from module import sub1, sub2
load sub2
```
Same here, only what is mention as `from . import ....` is loaded with the module.
You also see that it reminds what has been previously loaded.

-----------------
```
>>> from module import sub1
load initialization
import whole module
load sub1
```

When importing `sub1` the whole module is loaded anyway. But not other sub packages.

-----------------
```
>>> from module import *
load initialization
import whole module
>>> initialization
<module 'module.initialization' from '/tmp/module/initialization.py'>
>>> sub1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'sub1' is not defined
```
You might have expected as I did to see an import of sub1 and sub2 but it's not the case, which means that this is discouraged only due to the namespace filling.


Additional note: If you do want to load a subpackage which as a too generic name you could also define an alias on import using 
`from module import subpackage as msub`
This can save you from namespace conflicts.


In conclusion, the proposal of @sebalix is perfectly fine and has not impact other than having one different entry in the name space.


Some references:

https://docs.python.org/3/reference/import.html
https://docs.python.org/3/library/functions.html#__import__
https://stackoverflow.com/questions/12229580/python-importing-a-sub-package-or-sub-module